### PR TITLE
Comparable version representation

### PR DESCRIPTION
**Pull request by ebikt**
Ascii representations of debian versions can be comparend using simple ASCII dictionary ordering to get result of comparison of original version. Intended usage is to store ASCII representation in separate column in SQL table of packages and then use ORDER BY to obtain the package with the highest version.

### CODE REVIEW

**Comment by paultag** [comment]
@stapelberg mind doing a review? This is changing quite a bit, and you wrote this code originally.

File: version/version.go [hunk 1]
```diff
@@ -199,6 +397,9 @@ func parseInto(result *Version, input string) error {
 	}
 
 	colon := strings.Index(trimmed, ":")
+	if (colon >= 9) {
+		return fmt.Errorf("epoch too big or not an integer")
```
Comment (paultag) [inline version/version.go]: The epoch can be any unsigned integer; a valid version can be `10000000000:1.0-1`

**Comment by paultag** [comment]
This also may not help you, but there's a `postgresql-11-debversion` field for PostgreSQL that we use throughout Debian.

```
Description: Debian version number type for PostgreSQL
 Debian version numbers, used to version Debian binary and source
 packages, have a defined format, including specifications for how
 versions should be compared in order to sort them.  This package
 implements a "debversion" type to represent Debian version numbers
 within the PostgreSQL database.  This also includes operators for
 version comparison and index operator classes for creating indexes on
 the debversion type.
 .
 Version comparison uses the algorithm used by the Debian package
 manager, dpkg, using the implementation from libapt-pkg.  This means
 that columns in tables using the debversion type may be sorted and
 compared correctly using the same logic as "dpkg --compare-versions".
 It is also possible to create indexes on these columns.
 .
 postgresql-debversion implements the following features:
 .
  * The "debversion" type (internally derived from the "text" type)
  * A full set of operators for version comparison (< <= = <> >= >)
    including commutator and negator optimisation hints
  * Operator classes for btree and hash indexes
  * The aggregate functions min() and max()
```

File: version/version.go (hunk 1 above)
Comment (ebikt) [reply version/version.go]: Such epoch is valid, but not supported. It does not fit into 32bit integer, so most parsers would have problems (comparing it) anyways.

**Comment by ebikt** [comment]
> This also may not help you, but there's a `postgresql-11-debversion` field for PostgreSQL that we use throughout Debian.

Indeed, I was using this with some old version MySQL server. Having precomputed value for comparisons can be beneficial also for other uses (e.g., when sorting large number of versions, because byte-string comparison is much faster than interpreting debian version for every comparison.).

**Comment by paultag** [comment]
> Indeed, I was using this with some old version MySQL server. Having precomputed value for comparisons can be beneficial also for other uses (e.g., when sorting large number of versions, because byte-string comparison is much faster than interpreting debian version for every comparison.).

The PostgreSQL version here is index-able and is quite fast - we use it in production for the entire Debian archive :+1:

**Comment by paultag** [comment]
The more I think about this changeset, the more that I think this could be a separate package that takes a `debian/version.Version` and produces a string for SQL use. I don't know if I'm quite sold on this change to the version code, but let me read through the changes and think critically about it a bit more.

**Comment by ebikt** [comment]
I'm OK with the decision that it should be separate package. I wrote this code so maybe I should take responsibility for it in my own repository.

**Comment by paultag** [comment]
conflicts, we kinda petered out on this one, thanks for the changes!

**Pull request closed without merging**